# Testing JAAT package

Executed in Google Collab with A100 GPU.

## Load packages

In [2]:
# Install the package if not already installed
!pip install JAAT

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 69.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 166.4/166.4 MB 773.2 kB/s eta 0:00:00
  Created wheel for pure-predict: filename=pure_predict-0.0.4-py3-none-any.whl size=65388 sha256=b44ab74f8bd729b86146bdeaa846e7b8dbf7d8c84f711200be0d3a9588aa941f
  Stored in directory: /root/.cache/pip/wheels/da/b1/19/ade413ea97eb211d5cc4f5ed471d175703cf0ee21f1dd1f322
  Created wheel for swifter: filename=swifter-1.4.0-py3-none-any.whl size=16505 sha256=a7e2f51ca96a7a67ecb6edf6a24b4f1efc2a3147d150493730d127fe772ec3f7
  Stored in directory: /root/.cache/pip/wheels/d9/31/ff/ff51141a088571a9f672449e5aad5ea8bb35ca5d95ba135f30
Successfully built pure-predict swifter


In [14]:
# Import JAAT package
from JAAT import JAAT

# Load functions
tm = JAAT.TaskMatch()
sm = JAAT.SkillMatch()
tt = JAAT.TitleMatch()
fe = JAAT.FirmExtract()
we = JAAT.WageExtract()

INIT
Preparing embeddings...


Batches:   0%|          | 0/295 [00:00<?, ?it/s]

Setting up pipeline...


Device set to use cuda


Finished.
INIT
Preparing embeddings...


Batches:   0%|          | 0/28 [00:00<?, ?it/s]

Setting up pipeline...


Device set to use cuda


Finished.
INIT
Loading data...
Preparing embeddings...
Loading title models...


Device set to use cuda


Finished.


Device set to use cuda
Device set to use cuda
Device set to use cuda
Device set to use cuda


In [5]:
# Import other packages
import kagglehub
import pandas as pd
pd.set_option('display.max_colwidth', None)
import os

## Load sample job description publically availible data from Kaggle

[Data link](https://www.kaggle.com/datasets/kshitizregmi/jobs-and-job-description/data)

In [6]:
# Download data
path = kagglehub.dataset_download("kshitizregmi/jobs-and-job-description")

print("Path to dataset files:", path)

100%|██████████| 1.48M/1.48M [00:01<00:00, 1.40MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/kshitizregmi/jobs-and-job-description/versions/2


In [7]:
# Check file name
os.listdir(path)

['job_title_des.csv']

In [8]:
# Load data
df = pd.read_csv(os.path.join(path, "job_title_des.csv"))
df.head()

Unnamed: 0             Job Title  \
0           0     Flutter Developer   
1           1      Django Developer   
2           2      Machine Learning   
3           3         iOS Developer   
4           4  Full Stack Developer   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Job Description  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    

In [23]:
# Inspect data
df.info()
df.head()
df.sample(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2277 entries, 0 to 2276
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Unnamed: 0       2277 non-null   int64 
 1   Job Title        2277 non-null   object
 2   Job Description  2277 non-null   object
dtypes: int64(1), object(2)
memory usage: 53.5+ KB


Unnamed: 0         Job Title  \
1992        2093     iOS Developer   
401          414     PHP Developer   
96            97  Machine Learning   
133          134   DevOps Engineer   
1298        1357     PHP Developer   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

In [9]:
# Drop redundunt index column
df = df.drop(columns=["Unnamed: 0"])
df.head()

Job Title  \
0     Flutter Developer   
1      Django Developer   
2      Machine Learning   
3         iOS Developer   
4  Full Stack Developer   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Job Description  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

Job titles column seems to be inaccurate in some cases, therefore it is probably not a realible sourse of validation.

In [11]:
# Print few job descriptions
print(df['Job Description'][0])

We are looking for hire experts flutter developer. So you are eligible this post then apply your resume.
Job Types: Full-time, Part-time
Salary: ₹20,000.00 - ₹40,000.00 per month
Benefits:
Flexible schedule
Food allowance
Schedule:
Day shift
Supplemental Pay:
Joining bonus
Overtime pay
Experience:
total work: 1 year (Preferred)
Housing rent subsidy:
Yes
Industry:
Software Development
Work Remotely:
Temporarily due to COVID-19


In [12]:
print(df['Job Description'][4])

job responsibility full stack engineer – react role make impact petsmart transforming engineering team meet need rapidly changing retail environment role foundational helping build craft petsmart ’ prowess making react j native key framework across mobile web property working across web mobile application part building amazing mobile first customer experience impact customer store digital channel goal responsibility include act full stack developer across petsmart ’ various web mobile solution full stack engineer help build petsmart ’ competency using react j native across mobile web property partner web engineering team strategy share component across application best practice new advancement react community work product ux team review design concept offer suggestion decrease complexity maintaining spirit experience proactively make recommendation technical enhancement better performance scalability maintainability accountable delivering agile environment creates/oversees completion t

In [37]:
print(df['Job Description'][2])

Data Scientist (Contractor)

Bangalore, IN

Responsibilities

We are looking for a capable data scientist to join the Analytics team, reporting locally in India Bangalore. This person’s responsibilities include research, design and development of Machine Learning and Deep Learning algorithms to tackle a variety of Fraud oriented challenges. The data scientist will work closely with software engineers and program managers to deliver end-to-end products, including: data collection in big scale and analysis, exploring different algorithmic approaches, model development, assessment and validation – all the way through production.

Qualifications

At least 3 years of hands-on development of complex Machine Learning models using modern frameworks and tools, ideally Python based.
Solid understanding of statistics and applied mathematics
Creative thinker with a proven ability to tackle open problems and apply non-trivial solutions.
Experience in software development using Python, Java or a sim

The format of scraped job descriptions text is not standardized which reflects real-life data, which is a plus for testing.

## Title Exctraction

In [27]:
# Save first job description and title
example_text = df.loc[0, "Job Description"]
example_title = df.loc[0, "Job Title"]

In [28]:
example_text

'We are looking for hire experts flutter developer. So you are eligible this post then apply your resume.\nJob Types: Full-time, Part-time\nSalary: ₹20,000.00 - ₹40,000.00 per month\nBenefits:\nFlexible schedule\nFood allowance\nSchedule:\nDay shift\nSupplemental Pay:\nJoining bonus\nOvertime pay\nExperience:\ntotal work: 1 year (Preferred)\nHousing rent subsidy:\nYes\nIndustry:\nSoftware Development\nWork Remotely:\nTemporarily due to COVID-19'

In [29]:
example_title

'Flutter Developer'

In [30]:
# Title Exctraction
print("TITLE:", tt.get_title(example_text))


Extracting title values...


  0%|          | 0/1 [00:00<?, ?it/s]

Extracting title features...


  0%|          | 0/1 [00:00<?, ?it/s]

Matching titles to codes...


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

TITLE: [('13-1071.00', 0.872, -0.9, 'none')]


Does not look like a title...

It probably means:

*   '13-1071.00':
SOC / ONET occupation code.
Code 15-1243.00 corresponds to Human Resources Specialist in ONET.

*   0.872:
Semantic match score between your job title string and the SOC code’s canonical title text.

*   -0.9:
Output of their title_value regression model (some internal “value” score, not a title).

*   'none':
Output of the multi-label title_feature model: which title features are present. In this example, it predicted no special features.

This might be good or might be bad... Standartization is good in general, but how do we walk it back to output the text of the job title?

Also, Flutter Developer and DHuman Resources Specialist are different things... Is it due to text extraction model being wrong, match to occupation model being wrong, or me google wrong occupation code vintage?


Let's try for the data scientist title...

In [31]:
example_text = df.loc[2, "Job Description"]
example_text

'Data Scientist (Contractor)\n\nBangalore, IN\n\nResponsibilities\n\nWe are looking for a capable data scientist to join the Analytics team, reporting locally in India Bangalore. This person’s responsibilities include research, design and development of Machine Learning and Deep Learning algorithms to tackle a variety of Fraud oriented challenges. The data scientist will work closely with software engineers and program managers to deliver end-to-end products, including: data collection in big scale and analysis, exploring different algorithmic approaches, model development, assessment and validation – all the way through production.\n\nQualifications\n\nAt least 3 years of hands-on development of complex Machine Learning models using modern frameworks and tools, ideally Python based.\nSolid understanding of statistics and applied mathematics\nCreative thinker with a proven ability to tackle open problems and apply non-trivial solutions.\nExperience in software development using Python,

In [32]:
# Title Exctraction
print("TITLE:", tt.get_title(example_text))


Extracting title values...


  0%|          | 0/1 [00:00<?, ?it/s]

Extracting title features...


  0%|          | 0/1 [00:00<?, ?it/s]

Matching titles to codes...


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

TITLE: [('15-2051.00', 0.932, -0.3, 'none')]


15-2051.00 is the code for Data Scientist. Good!

Let's try full stack engineer...

In [38]:
example_text = df.loc[4, "Job Description"]
print("TITLE:", tt.get_title(example_text))


Extracting title values...


  0%|          | 0/1 [00:00<?, ?it/s]

Extracting title features...


  0%|          | 0/1 [00:00<?, ?it/s]

Matching titles to codes...


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

TITLE: [('15-1254.00', 0.886, -1.0, 'none')]


15-1254.00 are Web Developers...

And if I give it a title?

In [39]:
example_title= df.loc[4, "Job Title"]
print("TITLE:", tt.get_title(example_title))

Extracting title values...


  0%|          | 0/1 [00:00<?, ?it/s]

Extracting title features...


  0%|          | 0/1 [00:00<?, ?it/s]

Matching titles to codes...


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

TITLE: [('15-1252.00', 1.0, 0.0, 'none')]


Better. 15-1252.00 is Software Developer.

Let me try using the data package creators used in their deme. Did they extract job title from text? Or used already extracted title?

In [44]:
from operator import itemgetter


data = pd.read_excel("https://raw.githubusercontent.com/Job-Ad-Research-at-QSB-LUC/JAAT/main/data/demo/JobAdsData2022_OSF.xlsx")
data['text'] = data['text'].str.strip()
data.text = data.text.replace(r'\n','. ', regex=True)
data.text = data.text.apply(lambda x: x.lower() if isinstance(x, str) else x)
# We will work with the columns 'job_title' and 'text' (a concatenation of all text)
data.head(1)

id     onet_id           job_title                   company  \
0   1  43-4131.00  SBA Loan Associate  Bankers Healthcare Group   

          job_location salary  \
0  Fort Lauderdale, FL    NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     text  \
0  assist business development officer with sba 7(a) loan originations. assist business development officer with updated checklist for required loan origination documentation. provide weekly project summary detailing progress with each file. gather required documentation and organize for business development officer. document activity using our crm for business development officer. assist with marketing efforts of the business development officer. execute on established checklists and communicate effectively with customers and team members. work with team members to establish excellent customer service experience for all clients a successful candidate is a motivated administrative professional who is passionate about working in a growing and goal orientated environment. you excel at organization and being proactive, with a sense of urgency to ensure all agendas and timelines are completed in a timely manner. you are strong at multi-tasking and thrive in a fast-paced environment where you can work to assist team members in reaching company goals. this is a great opportunity to learn about the business and financial services industry.. customer service experience. administrative assistant prior work experience . intuitive, analytical and problem-solving skills are essential. sba 7(a) program experience is preferred. thri

Repeat what they did in demo.

In [48]:
data.job_title.to_list()

['SBA Loan Associate',
 'Credit Analyst',
 'Scientist, Lead Discovery',
 'Maintenance Director',
 'Engineer I',
 'Senior Research Scientist',
 'Family Physician',
 'Mortgate Underwriter',
 'Solar Electrical Project Engineer 2021',
 'Loan Delivery Specialist',
 'Medical Affairs Scientist',
 'Loan Originator Assistant',
 'Commissioning Technician I',
 'Mechanical Engineer',
 'Closing Processor III',
 'Facilities',
 'Maintenance Technician',
 'Facilities Maintenance Technician',
 'Equipment and Process Technician',
 'Maintenance Electrician',
 'Field Technician',
 'Maintenance Technician',
 'Field Operations Engineer III',
 'Inside Plant (ISP) Engineer I',
 'Instrument Technician',
 'Measurement Technician',
 'Refinery Operator',
 'G&P Project Engineer II',
 'Real Estate Processor',
 'Director, Process Engineering',
 'HVAC Technician',
 'Mechanic III',
 'Operating Engineer',
 'Chief Plant Engineer',
 'Medical Laboratory Scientist',
 'Field Service Technician',
 'Maintenance TEchnician',
 

In [45]:
data['TitleMatch'] = tt.get_title(data.job_title.to_list())
data['TM_onet_id'] = data.TitleMatch.apply(itemgetter(1)).astype(str)

Extracting title values...


You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset


  0%|          | 0/990 [00:00<?, ?it/s]

Extracting title features...


  0%|          | 0/62 [00:00<?, ?it/s]

Matching titles to codes...


Batches:   0%|          | 0/31 [00:00<?, ?it/s]

In [46]:
data.head()

id     onet_id                  job_title  \
0   1  43-4131.00         SBA Loan Associate   
1   2  43-4131.00             Credit Analyst   
2   3  19-1042.00  Scientist, Lead Discovery   
3   4  49-9071.00       Maintenance Director   
4   5  17-2112.00                 Engineer I   

                            company         job_location salary  \
0          Bankers Healthcare Group  Fort Lauderdale, FL    NaN   
1  First American Equipment Finance        Rochester, NY    NaN   
2         Accent Therapeutics, Inc.        Lexington, MA    NaN   
3           CWS Apartment Homes LLC           Austin, TX    NaN   
4          Freese and Nichols, Inc.          Atlanta, GA    NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               

### Conclusion 1

TitleMatch from JAAT does not extract job title from job description. It just turns it to ONET code, you need to extract job title yourself or have it in the data as a column to begin with.

According to the demo:
*   TitleMatch - Match Job Titles to O-NET Occupation Codes.
*   FirmNER - Extracts Candidate Firm Names from Job Ad (or other) Text.
*   TaskMatch - Match Job Ad Text to O-NET Task IDs
*   CREAM - Develop custom classifiers with embeddings by augmenting a human curated list of 'rules'. Demo uses Education Requirements.